strawberryfields/backends/bosonicbackend/bosoniccircuit.py

```diff
@@ -0,0 +1,833 @@
+        self.to_xp = to_xp(self.nlen)
+        self.from_xp = from_xp(self.nlen)
+        self.active.append(self.nlen - 1)
+
+        # Weights are set equal to each other and normalized
+        vac_weights = np.array([1 / num_gauss for i in range(num_gauss)], dtype=complex)
+        # New mode means and covs set to vacuum
+        vac_means = np.zeros((num_gauss, 2 * num_modes)).tolist()
+        vac_covs = [
+            ((self.hbar / 2) * np.identity(2 * num_modes)).tolist() for i in range(num_gauss)
+        ]
+
+        # Find all possible combinations of means and combs of the
+        # Gaussians between the modes.
+        mean_combs = it.product(self.means.tolist(), vac_means)
+        cov_combs = it.product(self.covs.tolist(), vac_covs)
+
+        # Tensor product of the weights.
+        weights = np.kron(self.weights, vac_weights)
+        # De-nest the means iterator.
+        means = np.array([[a for b in tup for a in b] for tup in mean_combs])
+        # Stack covs appropriately.
+        covs = np.array([block_diag(*tup) for tup in cov_combs])
+
+        self.weights = weights
+        self.means = means
+        self.covs = covs
+
+    def del_mode(self, modes):
+        r"""Delete modes from the circuit.
+
+        Args:
+            modes (int or list): modes to be deleted.
+        """
+        if isinstance(modes, int):
+            modes = [modes]
+
+        for mode in modes:
+            if self.active[mode] is None:
+                raise ValueError("Cannot delete mode, mode does not exist")
+
+            self.loss(0.0, mode)
+            self.active[mode] = None
+
+    def reset(self, num_subsystems=None, num_weights=1):
+        """Reset the simulation state.
+
+        Args:
+            num_subsystems (int, optional): Sets the number of modes in the reset
+                circuit. ``None`` means unchanged.
+            num_weights (int): Sets the number of gaussians per mode in the
+                superposition.
+        """
+        if num_subsystems is not None:
+            if not isinstance(num_subsystems, int):
+                raise ValueError("Number of modes must be an integer")
+            self.nlen = num_subsystems
+
+        if not isinstance(num_weights, int):
+            raise ValueError("Number of weights must be an integer")
+
+        self.active = list(np.arange(self.nlen, dtype=int))
+        # Mode index permutation list to and from XP ordering.
+        self.to_xp = to_xp(self.nlen)
+        self.from_xp = from_xp(self.nlen)
+
+        self.weights = np.ones(w_shape(self.nlen, num_weights), dtype=complex)
+        self.weights = self.weights / (num_weights ** self.nlen)
+
+        self.means = np.zeros(m_shape(self.nlen, num_weights), dtype=complex)
+        id_covs = [np.identity(2 * self.nlen, dtype=complex) for i in range(len(self.weights))]
+        self.covs = np.array(id_covs)
+
+    def get_modes(self):
+        r"""Return the modes currently active."""
```

Comment:
What is an active mode? What is an inactive mode? Maybe worth describing when a mode can be `None` in `self.active`

Reply:
I've clarified now.